### ISSUE TITLE
Apple Silicon createdump fails to find libmscordaccore.dylib (on cross/official builds)

### ISSUE THREAD

**Issue by sdmaclea** [dotnet/runtime#50932]
https://dev.azure.com/dnceng/public/_build/results?buildId=1077056&view=ms.vss-test-web.build-test-results-tab

```console
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at MyStruct180`1.IBase2<class BaseClass0,class BaseClass0>.Method7[M0]()
   at Framework.MethodCallingTest()
   at Framework.Main()
dyld: Library not loaded: @rpath/libmscordaccore.dylib
  Referenced from: /private/tmp/helix/working/B0A509A9/p/createdump
  Reason: image not found
/private/tmp/helix/working/B0A509A9/w/9D400930/e/Loader/classloader/TypeGeneratorTests/TypeGeneratorTest130/Generated130/Generated130.sh: line 377:  2764 Abort trap: 6           (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
```

```json
{
"HelixJobId": "b27d2a55-2b7f-4e96-b3a4-29db673eb82c",
"HelixWorkItemName": "Loader.classloader.TypeGeneratorTests.TypeGeneratorTest0-299"
}
```

**Comment by sdmaclea** [comment]
I haven't tried `createdump` locally with a custom directory path.  Perhaps that is the issue.

**Comment by sdmaclea** [comment]
My local M1 build succeeds

```console
stmaclea@stmaclea-arm64 runtime % COMPlus_DbgEnableElfDumpOnMacOS=1 artifacts/tests/coreclr/OSX.arm64.Checked/Tests/Core_Root/createdump
createdump [options] pid
-f, --name - dump path and file name. The default is '/tmp/coredump.%p'. These specifiers are substituted with following values:
   %p  PID of dumped process.
   %e  The process executable filename.
   %h  Hostname return by gethostname().
   %t  Time of dump, expressed as seconds since the Epoch, 1970-01-01 00:00:00 +0000 (UTC).
-n, --normal - create minidump.
-h, --withheap - create minidump with heap (default).
-t, --triage - create triage minidump.
-u, --full - create full core dump.
-d, --diag - enable diagnostic messages.
```

The helix payload on my local M1 machine fails
```console
stmaclea@stmaclea-arm64 runtime % COMPlus_DbgEnableElfDumpOnMacOS=1 ~/50932/createdump
dyld: Library not loaded: @rpath/libmscordaccore.dylib
  Referenced from: /Users/stmaclea/50932/createdump
  Reason: image not found
zsh: abort      COMPlus_DbgEnableElfDumpOnMacOS=1 ~/50932/createdump
```

**Comment by sdmaclea** [comment]
Also broken in preview2 build
```console
stmaclea@stmaclea-arm64 runtime % find /usr/local/share/dotnet -name createdump
/usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.0-preview.2.21154.6/createdump
stmaclea@stmaclea-arm64 runtime % /usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.0-preview.2.21154.6/createdump
dyld: Library not loaded: @rpath/libmscordaccore.dylib
  Referenced from: /usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.0-preview.2.21154.6/createdump
  Reason: image not found
zsh: abort      
stmaclea@stmaclea-arm64 runtime % dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.100-preview.2.21155.3
 Commit:    1a9103db2d

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  11.3
 OS Platform: Darwin
 RID:         osx.11.0-arm64
 Base Path:   /usr/local/share/dotnet/sdk/6.0.100-preview.2.21155.3/

Host (useful for support):
  Version: 6.0.0-preview.2.21154.6
  Commit:  3eaf1f316b

.NET SDKs installed:
  6.0.100-preview.1.21103.13 [/usr/local/share/dotnet/sdk]
  6.0.100-preview.2.21155.3 [/usr/local/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.0-preview.1.21103.6 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.2.21154.6 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.0-preview.2.21154.6 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```

**Comment by k15tfu** [comment]
Hi! I can reproduce this issue with preview3 `dotnet-sdk-6.0.100-preview.3.21202.5-osx-arm64` and with [preview3/daily](https://aka.ms/dotnet/6.0.1XX-preview3/daily/dotnet-sdk-osx-arm64.tar.gz) `dotnet-sdk-6.0.100-preview.3.21207.34-osx-arm64`
```
% dotnet-sdk-6.0.100-preview.3.21202.5-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump
dyld: Library not loaded: @rpath/libmscordaccore.dylib
  Referenced from: .../dotnet-sdk-6.0.100-preview.3.21202.5-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump
  Reason: image not found
zsh: abort
% dotnet-sdk-6.0.100-preview.3.21207.34-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump
dyld: Library not loaded: @rpath/libmscordaccore.dylib
  Referenced from: .../dotnet-sdk-6.0.100-preview.3.21207.34-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump
  Reason: image not found
zsh: abort
```

but this issue seems to be fixed in [preview4/daily](https://aka.ms/dotnet/6.0.1XX-preview4/daily/dotnet-sdk-osx-arm64.tar.gz) `dotnet-sdk-6.0.100-preview.4.21223.2-osx-arm64`:
```
% dotnet-sdk-6.0.100-preview.4.21223.2-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.4.21222.10//createdump
createdump [options] pid
-f, --name - dump path and file name. The default is '/tmp/coredump.%p'. These specifiers are substituted with following values:
   %p  PID of dumped process.
   %e  The process executable filename.
   %h  Hostname return by gethostname().
   %t  Time of dump, expressed as seconds since the Epoch, 1970-01-01 00:00:00 +0000 (UTC).
-n, --normal - create minidump.
-h, --withheap - create minidump with heap (default).
-t, --triage - create triage minidump.
-u, --full - create full core dump.
-d, --diag - enable diagnostic messages.
-vd, --verbose - enable verbose diagnostic messages.
```

Not sure what changed, its dependencies are the same (thx https://github.com/dotnet/runtime/pull/45978):
```
% otool -L dotnet-sdk-6.0.100-preview.3.21202.5-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump
dotnet-sdk-6.0.100-preview.3.21202.5-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump:
        @rpath/libmscordaccore.dylib (compatibility version 0.0.0, current version 0.0.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 904.4.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.0.0)
% otool -L dotnet-sdk-6.0.100-preview.3.21207.34-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump
dotnet-sdk-6.0.100-preview.3.21207.34-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump:
        @rpath/libmscordaccore.dylib (compatibility version 0.0.0, current version 0.0.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 904.4.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.0.0)
% otool -L dotnet-sdk-6.0.100-preview.4.21223.2-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.4.21222.10/createdump
dotnet-sdk-6.0.100-preview.4.21223.2-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.4.21222.10/createdump:
        @rpath/libmscordaccore.dylib (compatibility version 0.0.0, current version 0.0.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 904.4.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.0.0)
(and for osx-x64)
% otool -L dotnet-sdk-6.0.100-preview.3.21202.5-osx-x64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump
dotnet-sdk-6.0.100-preview.3.21202.5-osx-x64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump:
        @rpath/libmscordaccore.dylib (compatibility version 0.0.0, current version 0.0.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 904.4.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
```

but I found that `LC_RPATH` are different:
```
% otool -l dotnet-sdk-6.0.100-preview.3.21202.5-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump | grep -B1 -A2 RPATH
% otool -l dotnet-sdk-6.0.100-preview.3.21207.34-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump | grep -B1 -A2 RPATH
% otool -l dotnet-sdk-6.0.100-preview.4.21223.2-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.4.21222.10/createdump | grep -B1 -A2 RPATH
Load command 16
          cmd LC_RPATH
      cmdsize 32
         path @loader_path (offset 12)
(and for osx-x64)
% otool -l dotnet-sdk-6.0.100-preview.3.21202.5-osx-x64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/createdump | grep -B1 -A2 RPATH
Load command 15
          cmd LC_RPATH
      cmdsize 32
         path @loader_path (offset 12)
```

Now I faced with similar issue when loading `libmscordbi.dylib`:
```
dlopen(.../dotnet-sdk-6.0.100-preview.3.21202.5-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/libmscordbi.dylib, 1): Library not loaded: @rpath/libmscordaccore.dylib
  Referenced from: .../dotnet-sdk-6.0.100-preview.3.21202.5-osx-arm64/shared/Microsoft.NETCore.App/6.0.0-preview.3.21201.4/libmscordbi.dylib
  Reason: image not found
```

@sdmaclea What was the reason behind this change https://github.com/dotnet/runtime/commit/60b6af1beaf6f973d80cf28d78f3284e2df342c7#diff-befe4aff86a59de86de3995195e0b3b8b84be87e264cc2e2ab709230df916f76R7-R8?

The same configuration is used in createdump/CMakeLists.txt, and if we revert this change in mscordbi/CMakeLists.txt the problem with loading mscordbi will be fixed as well. Here is PR: https://github.com/dotnet/runtime/pull/51751.

**Comment by sdmaclea** [comment]
> What was the reason behind this change?

When built with the latest XCode the library search was not successful in early macOS 11.x versions.  The docs were suggesting the method we use for x64 was deprecated and would eventually be removed. I assumed Apple had chosen to remove it with the move to macOS 11.  It is possible it was just working around a bug in early macOS 11.

**Comment by sdmaclea** [comment]
Fixed by #51751
